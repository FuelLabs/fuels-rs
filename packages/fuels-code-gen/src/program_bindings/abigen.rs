use std::{collections::HashSet, path::PathBuf};

pub use abigen_target::{AbigenTarget, ProgramType};
use fuel_abi_types::abi::full_program::FullTypeDeclaration;
use inflector::Inflector;
use itertools::Itertools;
use proc_macro2::TokenStream;
use quote::quote;
use regex::Regex;

use crate::{
    error::Result,
    program_bindings::{
        abigen::{abigen_target::ParsedAbigenTarget, bindings::generate_bindings},
        custom_types::generate_types,
        generated_code::GeneratedCode,
    },
    utils::ident,
};

mod abigen_target;
mod bindings;
mod configurables;
mod logs;

pub struct Abigen;

impl Abigen {
    /// Generate code which can be used to interact with the underlying
    /// contract, script or predicate in a type-safe manner.
    ///
    /// # Arguments
    ///
    /// * `targets`: `AbigenTargets` detailing which ABI to generate bindings
    /// for, and of what nature (Contract, Script or Predicate).
    /// * `no_std`: don't use the Rust std library.
    pub fn generate(targets: Vec<AbigenTarget>, no_std: bool) -> Result<TokenStream> {
        let parsed_targets = Self::parse_targets(targets)?;

        let generated_code = Self::generate_code(no_std, parsed_targets)?;

        let use_statements = generated_code.use_statements_for_uniquely_named_types();

        let code = if no_std {
            Self::wasm_paths_hotfix(&generated_code.code())
        } else {
            generated_code.code()
        };

        Ok(quote! {
            #code
            #use_statements
        })
    }
    fn wasm_paths_hotfix(code: &TokenStream) -> TokenStream {
        [
            (r#"::\s*std\s*::\s*string"#, "::alloc::string"),
            (r#"::\s*std\s*::\s*format"#, "::alloc::format"),
            (r#"::\s*std\s*::\s*vec"#, "::alloc::vec"),
            (r#"::\s*std\s*::\s*boxed"#, "::alloc::boxed"),
        ]
        .map(|(reg_expr_str, substitute)| (Regex::new(reg_expr_str).unwrap(), substitute))
        .into_iter()
        .fold(code.to_string(), |code, (regex, wasm_include)| {
            regex.replace_all(&code, wasm_include).to_string()
        })
        .parse()
        .expect("Wasm hotfix failed!")
    }

    fn generate_code(
        no_std: bool,
        parsed_targets: Vec<ParsedAbigenTarget>,
    ) -> Result<GeneratedCode> {
        let custom_types = Self::filter_custom_types(&parsed_targets);
        let shared_types = Self::filter_shared_types(custom_types);

        let bindings = Self::generate_all_bindings(parsed_targets, no_std, &shared_types)?;
        let shared_types = Self::generate_shared_types(shared_types, no_std)?;

        let mod_name = ident("abigen_bindings");
        Ok(shared_types.merge(bindings).wrap_in_mod(mod_name))
    }

    fn generate_all_bindings(
        parsed_targets: Vec<ParsedAbigenTarget>,
        no_std: bool,
        shared_types: &HashSet<FullTypeDeclaration>,
    ) -> Result<GeneratedCode> {
        parsed_targets
            .into_iter()
            .map(|target| Self::generate_binding(target, no_std, shared_types))
            .fold_ok(GeneratedCode::default(), |acc, generated_code| {
                acc.merge(generated_code)
            })
    }

    fn generate_binding(
        target: ParsedAbigenTarget,
        no_std: bool,
        shared_types: &HashSet<FullTypeDeclaration>,
    ) -> Result<GeneratedCode> {
        let mod_name = ident(&format!("{}_mod", &target.name.to_snake_case()));

        let recompile_trigger =
            Self::generate_macro_recompile_trigger(target.source.path.as_ref(), no_std);
        let types = generate_types(&target.source.abi.types, shared_types, no_std)?;
        let bindings = generate_bindings(target, no_std)?;
        Ok(recompile_trigger
            .merge(types)
            .merge(bindings)
            .wrap_in_mod(mod_name))
    }

    /// Any changes to the file pointed to by `path` will cause the reevaluation of the current
    /// procedural macro. This is a hack until https://github.com/rust-lang/rust/issues/99515
    /// lands.
    fn generate_macro_recompile_trigger(path: Option<&PathBuf>, no_std: bool) -> GeneratedCode {
        let code = path
            .as_ref()
            .map(|path| {
                let stringified_path = path.display().to_string();
                quote! {
                    const _: &[u8] = include_bytes!(#stringified_path);
                }
            })
            .unwrap_or_default();
        GeneratedCode::new(code, Default::default(), no_std)
    }

    fn parse_targets(targets: Vec<AbigenTarget>) -> Result<Vec<ParsedAbigenTarget>> {
        targets
            .into_iter()
            .map(|target| target.try_into())
            .collect()
    }

    fn generate_shared_types(
        shared_types: HashSet<FullTypeDeclaration>,
        no_std: bool,
    ) -> Result<GeneratedCode> {
        let types = generate_types(&shared_types, &HashSet::default(), no_std)?;

        if types.is_empty() {
            Ok(Default::default())
        } else {
            let mod_name = ident("shared_types");
            Ok(types.wrap_in_mod(mod_name))
        }
    }

    fn filter_custom_types(
        all_types: &[ParsedAbigenTarget],
    ) -> impl Iterator<Item = &FullTypeDeclaration> {
        all_types
            .iter()
            .flat_map(|target| &target.source.abi.types)
            .filter(|ttype| ttype.is_custom_type())
    }

    /// A type is considered "shared" if it appears at least twice in
    /// `all_custom_types`.
    ///
    /// # Arguments
    ///
    /// * `all_custom_types`: types from all ABIs whose bindings are being
    /// generated.
    fn filter_shared_types<'a>(
        all_custom_types: impl IntoIterator<Item = &'a FullTypeDeclaration>,
    ) -> HashSet<FullTypeDeclaration> {
        all_custom_types.into_iter().duplicates().cloned().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn correctly_determines_shared_types() {
        let types = ["type_0", "type_1", "type_0"].map(|type_field| FullTypeDeclaration {
            type_field: type_field.to_string(),
            components: vec![],
            type_parameters: vec![],
        });

        let shared_types = Abigen::filter_shared_types(&types);

        assert_eq!(shared_types, HashSet::from([types[0].clone()]))
    }
}
