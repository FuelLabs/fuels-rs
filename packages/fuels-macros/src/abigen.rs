mod code_gen;
mod parsing;

pub(crate) use code_gen::{Abigen, AbigenTarget, ProgramType, TypePath};
pub(crate) use parsing::MacroAbigenTargets;
