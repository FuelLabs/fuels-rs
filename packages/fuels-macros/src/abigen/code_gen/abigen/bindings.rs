use std::collections::HashSet;

use crate::{
    abigen::code_gen::{
        abi_types::FullTypeDeclaration,
        abigen::{
            abigen_target::ParsedAbigenTarget,
            bindings::{
                contract::contract_bindings, predicate::predicate_bindings, script::script_bindings,
            },
            ProgramType,
        },
        generated_code::GeneratedCode,
    },
    error::Result,
    utils::ident,
};

mod contract;
mod function_generator;
mod predicate;
mod script;
mod utils;

pub(crate) fn generate_bindings(
    target: ParsedAbigenTarget,
    no_std: bool,
    shared_types: &HashSet<FullTypeDeclaration>,
) -> Result<GeneratedCode> {
    let bindings_generator = match target.program_type {
        ProgramType::Script => script_bindings,
        ProgramType::Contract => contract_bindings,
        ProgramType::Predicate => predicate_bindings,
    };

    let name = ident(&target.name);
    let abi = target.source;
    bindings_generator(&name, abi, no_std, shared_types)
}
