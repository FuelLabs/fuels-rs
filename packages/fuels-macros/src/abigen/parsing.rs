use fuels_code_gen::{AbigenTarget, ProgramType};
use syn::{
    parse::{Parse, ParseStream},
    Result,
};

use crate::parse_utils::{Command, UniqueNameValues};

impl From<MacroAbigenTargets> for Vec<AbigenTarget> {
    fn from(targets: MacroAbigenTargets) -> Self {
        targets.targets.into_iter().map(Into::into).collect()
    }
}

impl From<MacroAbigenTarget> for AbigenTarget {
    fn from(macro_target: MacroAbigenTarget) -> Self {
        AbigenTarget {
            name: macro_target.name,
            abi: macro_target.abi,
            program_type: macro_target.program_type,
        }
    }
}

// Although identical to `AbigenTarget` from fuels-core, due to the orphan rule
// we cannot implement Parse for the latter.
#[derive(Debug)]
pub(crate) struct MacroAbigenTarget {
    pub(crate) name: String,
    pub(crate) abi: String,
    pub(crate) program_type: ProgramType,
}

pub(crate) struct MacroAbigenTargets {
    targets: Vec<MacroAbigenTarget>,
}

impl Parse for MacroAbigenTargets {
    fn parse(input: ParseStream) -> Result<Self> {
        let targets = Command::parse_multiple(input)?
            .into_iter()
            .map(MacroAbigenTarget::new)
            .collect::<Result<_>>()?;

        Ok(Self { targets })
    }
}

impl MacroAbigenTarget {
    pub fn new(command: Command) -> Result<Self> {
        let program_type = command.name.try_into()?;

        let name_values = UniqueNameValues::new(command.contents)?;
        name_values.validate_has_no_other_names(&["name", "abi"])?;

        let name = name_values.get_as_lit_str("name")?.value();
        let abi = name_values.get_as_lit_str("abi")?.value();

        Ok(Self {
            name,
            abi,
            program_type,
        })
    }
}
