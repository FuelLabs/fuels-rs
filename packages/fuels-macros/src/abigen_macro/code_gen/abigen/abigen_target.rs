use std::convert::TryFrom;

use crate::{
    abigen_macro::code_gen::{abi_types::FullProgramABI, source::Source},
    error::{Error, Result},
};

#[derive(Debug, Clone)]
pub struct AbigenTarget {
    pub name: String,
    pub abi: String,
    pub program_type: ProgramType,
}

pub(crate) struct ParsedAbigenTarget {
    pub name: String,
    pub source: FullProgramABI,
    pub program_type: ProgramType,
}

impl TryFrom<AbigenTarget> for ParsedAbigenTarget {
    type Error = Error;

    fn try_from(value: AbigenTarget) -> Result<Self> {
        Ok(Self {
            name: value.name,
            source: parse_program_abi(&value.abi)?,
            program_type: value.program_type,
        })
    }
}

fn parse_program_abi(abi_source: &str) -> Result<FullProgramABI> {
    let source = Source::parse(abi_source).expect("failed to parse JSON ABI");
    let json_abi_str = source.get().expect("failed to parse JSON ABI from string");
    FullProgramABI::from_json_abi(&json_abi_str)
}

#[derive(Debug, Clone, Copy)]
pub enum ProgramType {
    Script,
    Contract,
    Predicate,
}
