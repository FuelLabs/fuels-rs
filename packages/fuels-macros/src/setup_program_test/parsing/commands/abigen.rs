use std::convert::TryFrom;

use fuels_code_gen::ProgramType;
use proc_macro2::Span;
use syn::{Error, LitStr};

use crate::parse_utils::{Command, UniqueNameValues};

#[derive(Debug, Clone)]
pub(crate) struct TargetInfo {
    pub(crate) name: LitStr,
    pub(crate) project: LitStr,
    pub(crate) program_type: ProgramType,
}

impl TryFrom<Command> for TargetInfo {
    type Error = Error;

    fn try_from(command: Command) -> Result<Self, Self::Error> {
        let program_type = command.name.try_into()?;

        let name_values = UniqueNameValues::new(command.contents)?;
        name_values.validate_has_no_other_names(&["name", "project"])?;

        let name = name_values.get_as_lit_str("name")?.clone();
        let project = name_values.get_as_lit_str("project")?.clone();

        Ok(Self {
            name,
            project,
            program_type,
        })
    }
}

#[derive(Debug, Clone)]
pub(crate) struct AbigenCommand {
    pub(crate) span: Span,
    pub(crate) targets: Vec<TargetInfo>,
}

impl TryFrom<Command> for AbigenCommand {
    type Error = Error;

    fn try_from(command: Command) -> Result<Self, Self::Error> {
        let targets = command
            .contents
            .into_iter()
            .map(|meta| Command::new(meta).and_then(TargetInfo::try_from))
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self {
            span: command.name.span(),
            targets,
        })
    }
}
