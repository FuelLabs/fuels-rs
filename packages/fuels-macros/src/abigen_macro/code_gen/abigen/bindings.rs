use crate::abigen_macro::code_gen::abi_types::FullTypeDeclaration;
use crate::abigen_macro::code_gen::abigen::abigen_target::ParsedAbigenTarget;
use crate::abigen_macro::code_gen::abigen::bindings::contract::contract_bindings;
use crate::abigen_macro::code_gen::abigen::bindings::predicate::predicate_bindings;
use crate::abigen_macro::code_gen::abigen::bindings::script::script_bindings;
use crate::abigen_macro::code_gen::abigen::ProgramType;
use crate::abigen_macro::code_gen::generated_code::GeneratedCode;
use crate::utils::ident;
use std::collections::HashSet;

mod contract;
mod function_generator;
mod predicate;
mod script;
mod utils;

pub(crate) fn generate_bindings(
    target: ParsedAbigenTarget,
    no_std: bool,
    shared_types: &HashSet<FullTypeDeclaration>,
) -> crate::Result<GeneratedCode> {
    let bindings_generator = match target.program_type {
        ProgramType::Script => script_bindings,
        ProgramType::Contract => contract_bindings,
        ProgramType::Predicate => predicate_bindings,
    };

    let name = ident(&target.name);
    let abi = target.source;
    bindings_generator(&name, abi, no_std, shared_types)
}
