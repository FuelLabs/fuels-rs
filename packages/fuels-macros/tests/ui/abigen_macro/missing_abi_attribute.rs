use fuels_macros::abigen;

abigen!(Contract(name = "SomeName"));

fn main() {}
