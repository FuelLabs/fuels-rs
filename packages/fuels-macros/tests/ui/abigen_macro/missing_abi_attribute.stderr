error: Missing attribute 'abi'.
 --> tests/ui/abigen_macro/missing_abi_attribute.rs:3:18
  |
3 | abigen!(Contract(name = "SomeName"));
  |                  ^^^^
